File: app/src/main/java/com/kupay/kupay/util/ViewUtil.java
package com.kupay.kupay.util;

import android.app.Activity;
import android.content.Context;
import android.content.res.Resources;
import android.util.DisplayMetrics;

import java.lang.reflect.Method;

/**
 * Created by "iqos_jay@outlook.com" on 2018/7/6.
 */
public class ViewUtil {
    /**
     * 获取设备屏幕的宽度
     *
     * @param ctx 上下文
     * @return 屏幕的宽度
     */
    public static float getScreenWidth(Activity ctx) {
        Resources resources = ctx.getResources();
        DisplayMetrics dm = resources.getDisplayMetrics();
        return dm.widthPixels;
    }

    /**
     * 获取设备屏幕的高度
     *
     * @param ctx 上下文
     * @return 屏幕的高度
     */
    public static float getScreenHeight(Activity ctx) {
        Resources resources = ctx.getResources();
        DisplayMetrics dm = resources.getDisplayMetrics();
        return dm.heightPixels;
    }

    /**
     * 获取状态栏的高度
     *
     * @param ctx 上下文
     * @return 状态栏的高度
     */
    public static float getStatusBarHeight(Context ctx) {
        int statusBarHeight = -1;
        //获取status_bar_height资源的ID
        int resourceId = ctx.getResources().getIdentifier("status_bar_height", "dimen", "android");
        if (resourceId > 0) {
            //根据资源ID获取响应的尺寸值
            statusBarHeight = ctx.getResources().getDimensionPixelSize(resourceId);
        }
        return statusBarHeight;
    }

    /**
     * 根据手机的分辨率从 dp 的单位 转成为 px(像素)
     */
    public static int dip2px(Context context, float dpValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    /**
     * 根据手机的分辨率从 px(像素) 的单位 转成为 dp
     */
    public static int px2dip(Context context, float pxValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (pxValue / scale + 0.5f);
    }

    //获取是否存在NavigationBar
    private static boolean checkDeviceHasNavigationBar(Context context) {
        boolean hasNavigationBar = false;
        Resources rs = context.getResources();
        int id = rs.getIdentifier("config_showNavigationBar", "bool", "android");
        if (id > 0) {
            hasNavigationBar = rs.getBoolean(id);
        }
        try {
            Class systemPropertiesClass = Class.forName("android.os.SystemProperties");
            Method m = systemPropertiesClass.getMethod("get", String.class);
            String navBarOverride = (String) m.invoke(systemPropertiesClass, "qemu.hw.mainkeys");
            if ("1".equals(navBarOverride)) {
                hasNavigationBar = false;
            } else if ("0".equals(navBarOverride)) {
                hasNavigationBar = true;
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        return hasNavigationBar;
    }

    /**
     * 获取NavigationBar高度
     *
     * @param context
     * @return
     */
    public static int getNavigationBarHeight(Context context) {
        boolean hasNavigationBar = checkDeviceHasNavigationBar(context);
        if (!hasNavigationBar) return 0;
        Resources resources = context.getResources();
        int resourceId = resources.getIdentifier("navigation_bar_height", "dimen", "android");
        return resources.getDimensionPixelSize(resourceId);
    }
}

File: pi_framework/src/main/java/com/kuplay/pi_framework/Util/ViewUtil.kt
package com.kuplay.pi_framework.Util

import android.app.Activity
import android.content.Context

object ViewUtil {
    /**
     * 获取设备屏幕的宽度
     *
     * @param ctx 上下文
     * @return 屏幕的宽度
     */
    fun getScreenWidth(ctx: Activity): Float {
        val resources = ctx.resources
        val dm = resources.displayMetrics
        return dm.widthPixels.toFloat()
    }

    /**
     * 获取设备屏幕的高度
     *
     * @param ctx 上下文
     * @return 屏幕的高度
     */
    fun getScreenHeight(ctx: Activity): Float {
        val resources = ctx.resources
        val dm = resources.displayMetrics
        return dm.heightPixels.toFloat()
    }

    /**
     * 获取状态栏的高度
     *
     * @param ctx 上下文
     * @return 状态栏的高度
     */
    fun getStatusBarHeight(ctx: Context): Float {
        var statusBarHeight = -1
        //获取status_bar_height资源的ID
        val resourceId = ctx.resources.getIdentifier("status_bar_height", "dimen", "android")
        if (resourceId > 0) {
            //根据资源ID获取响应的尺寸值
            statusBarHeight = ctx.resources.getDimensionPixelSize(resourceId)
        }
        return statusBarHeight.toFloat()
    }

    /**
     * 根据手机的分辨率从 dp 的单位 转成为 px(像素)
     */
    fun dip2px(context: Context, dpValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (dpValue * scale + 0.5f).toInt()
    }

    /**
     * 根据手机的分辨率从 px(像素) 的单位 转成为 dp
     */
    fun px2dip(context: Context, pxValue: Float): Int {
        val scale = context.resources.displayMetrics.density
        return (pxValue / scale + 0.5f).toInt()
    }

    //获取是否存在NavigationBar
    private fun checkDeviceHasNavigationBar(context: Context): Boolean {
        var hasNavigationBar = false
        val rs = context.resources
        val id = rs.getIdentifier("config_showNavigationBar", "bool", "android")
        if (id > 0) {
            hasNavigationBar = rs.getBoolean(id)
        }
        try {
            val systemPropertiesClass = Class.forName("android.os.SystemProperties")
            val m = systemPropertiesClass.getMethod("get", String::class.java)
            val navBarOverride = m.invoke(systemPropertiesClass, "qemu.hw.mainkeys") as String
            if ("1" == navBarOverride) {
                hasNavigationBar = false
            } else if ("0" == navBarOverride) {
                hasNavigationBar = true
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return hasNavigationBar
    }

    /**
     * 获取NavigationBar高度
     *
     * @param context
     * @return
     */
    fun getNavigationBarHeight(context: Context): Int {
        val hasNavigationBar = checkDeviceHasNavigationBar(context)
        if (!hasNavigationBar) return 0
        val resources = context.resources
        val resourceId = resources.getIdentifier("navigation_bar_height", "dimen", "android")
        return resources.getDimensionPixelSize(resourceId)
    }
}